pkg/driver/driver.go

```diff
@@ -43,56 +47,43 @@ func (c RetryConnector) Connect(ctx context.Context) (driver.Conn, error) {
 				telemetry.UpdateCurrentDbConnErr(err)
 
 				if lastErr == nil || err.Error() != lastErr.Error() {
-					c.driver.Logger.Warnw("Can't connect to database. Retrying", zap.Error(err))
+					c.Logger.Warnw("Can't connect to database. Retrying", zap.Error(err))
 				}
 			},
 			OnSuccess: func(elapsed time.Duration, attempt uint64, _ error) {
 				telemetry.UpdateCurrentDbConnErr(nil)
 
 				if attempt > 0 {
-					c.driver.Logger.Infow("Reconnected to database",
+					c.Logger.Infow("Reconnected to database",
 						zap.Duration("after", elapsed), zap.Uint64("attempts", attempt+1))
 				}
 			},
 		},
 	), "can't connect to database")
+
+	if err == nil && c.InitConn != nil {
+		err = c.InitConn(ctx, conn)
+		if err != nil {
+			return nil, err
+		}
+	}
+
 	return conn, err
 }
 
 // Driver implements part of the driver.Connector interface.
 func (c RetryConnector) Driver() driver.Driver {
-	return c.driver
-}
-
-// Driver wraps a driver.Driver that also must implement driver.DriverContext with logging capabilities and provides our RetryConnector.
-type Driver struct {
-	ctxDriver
-	Logger *logging.Logger
-}
-
-// OpenConnector implements the DriverContext interface.
-func (d Driver) OpenConnector(name string) (driver.Connector, error) {
-	c, err := d.ctxDriver.OpenConnector(name)
-	if err != nil {
-		return nil, err
-	}
-
-	return &RetryConnector{
-		driver:    d,
-		Connector: c,
-	}, nil
+	return &ctxDriver{}
 }
```

Comment:
The `Driver()` method doesn't look right. If it would ever be used, the result would be useless.

Maybe keep the `Driver` type, store the `InitConn` callback there and use it for every connection created from that driver.

Reply:
> The `Driver()` method doesn't look right. If it would ever be used, the result would be useless.

[`PostgreSQL`](https://github.com/lib/pq/blob/master/connector.go#L36-L38) does exactly the same thing.

Reply:
Not really the same.

`pq.Driver` has a useful implementation of the `Driver` on a zero-initialized value: https://github.com/lib/pq/blob/3d613208bca2e74f2a20e04126ed30bcb5c4cc27/conn.go#L50-L58

`ctxDriver` on the other hand just embeds two interfaces. So if a method on a zero-initialized value is called, this will just result in a nil derefence.